server/src/main/java/org/elasticsearch/index/mapper/IndexTimeScriptContext.java

```diff
@@ -0,0 +1,256 @@
+import org.apache.lucene.index.DocValuesType;
+import org.apache.lucene.index.FieldInfo;
+import org.apache.lucene.index.FieldInfos;
+import org.apache.lucene.index.Fields;
+import org.apache.lucene.index.IndexOptions;
+import org.apache.lucene.index.IndexableField;
+import org.apache.lucene.index.LeafMetaData;
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.NumericDocValues;
+import org.apache.lucene.index.PointValues;
+import org.apache.lucene.index.SortedDocValues;
+import org.apache.lucene.index.SortedNumericDocValues;
+import org.apache.lucene.index.SortedSetDocValues;
+import org.apache.lucene.index.StoredFieldVisitor;
+import org.apache.lucene.index.Terms;
+import org.apache.lucene.index.memory.MemoryIndex;
+import org.apache.lucene.util.Bits;
+import org.elasticsearch.common.CheckedConsumer;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.index.fielddata.IndexFieldDataCache;
+import org.elasticsearch.indices.breaker.NoneCircuitBreakerService;
+import org.elasticsearch.search.lookup.SearchLookup;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+public class IndexTimeScriptContext {
+
+    public final SearchLookup searchLookup;
+    public final LeafReaderContext leafReaderContext;
+    public final ParseContext pc;
+    private final Map<String, MapperScript> mapperScripts = new HashMap<>();
+
+    public static void executePostParsePhases(MappingLookup lookup, ParseContext parseContext) throws IOException {
+        if (lookup.getPostParsePhases().isEmpty()) {
+            return;
+        }
+        IndexTimeScriptContext context = new IndexTimeScriptContext(lookup, parseContext);
+        context.executePostParse();
+    }
+
+    private IndexTimeScriptContext(MappingLookup mappingLookup, ParseContext pc) {
+        this.searchLookup = new SearchLookup(
+            mappingLookup::getFieldType,
+            (ft, s) -> ft.fielddataBuilder("", s).build(
+                new IndexFieldDataCache.None(),
+                new NoneCircuitBreakerService())
+        );
+        this.pc = pc;
+        LazyDocumentReader reader = new LazyDocumentReader(
+            pc.rootDoc(),
+            pc.sourceToParse().source(),
+            mappingLookup.getPostParsePhases().keySet());
+        this.leafReaderContext = reader.getContext();
+        mappingLookup.getPostParsePhases().forEach((k, c) -> mapperScripts.put(k, new MapperScript(c)));
+    }
+
+    private void executePostParse() throws IOException {
+        for (MapperScript ms : mapperScripts.values()) {
+            ms.script.accept(this);
+        }
+    }
+
+    private void executeFieldScript(String field) throws IOException {
+        assert mapperScripts.containsKey(field);
+        mapperScripts.get(field).script.accept(this);
+        mapperScripts.get(field).script = c -> {};
```

Comment:
This stops scripts being executed multiple times if they are referred to by other scripts

Reply:
It might be better to set a boolean. I expect there won't be a real speed difference and it'd probably be a bit easier to follow.

Reply:
++